File: year_2023/day_10/day10.py
from dataclasses import dataclass

import numpy as np
from matplotlib.path import Path

from tools import loader, parsers


@dataclass
class Cart:
    location: complex
    start: complex
    direction: complex = 0j

    def rotate(self, tile: str) -> None:
        match tile:
            case 'L' | '7':
                self.direction *= -1j if self.direction.real == 0 else 1j
            case 'J' | 'F':
                self.direction *= 1j if self.direction.real == 0 else -1j


class Pipe:
    """Code borrowed from 2018 day 13"""

    def __init__(self, data: list[str]) -> None:
        self.pipes = {}
        self.arr = np.genfromtxt(data, delimiter=1, dtype=str)

        for (i, j), tile in np.ndenumerate(self.arr):
            location = complex(i, j)
            if tile in '-|F7JLS':
                self.pipes[location] = tile
            if tile == 'S':
                self.cart = Cart(location, location)

        self.cart.direction = next(i for i in (1, -1, 1j, -1j)
                                   if self.cart.location + i in self.pipes)
        self.track = self.get_pipe_circle(self.cart)

    def get_pipe_circle(self, cart: Cart) -> list[complex]:
        track = []
        while True:
            track.append(cart.location)
            cart.location += cart.direction
            cart.rotate(self.pipes.get(cart.location))
            if cart.location == cart.start:
                break
        return track

    def part_1(self) -> int:
        """
        >>> print(Pipe(parsers.lines('test.txt')).part_1())
        8"""
        return len(self.track) // 2

    def part_2(self) -> int:
        """
        >>> print(Pipe(parsers.lines('test2.txt')).part_2())
        4
        >>> print(Pipe(parsers.lines('test3.txt')).part_2())
        8"""
        path = Path(np.array([(int(i.real), int(i.imag)) for i in self.track]))
        return sum(
            complex(*point) not in self.track and path.contains_point(point)
            for point, _ in np.ndenumerate(self.arr)
        )


print(Pipe(parsers.lines(loader.get())).part_1())  # 7086
print(Pipe(parsers.lines(loader.get())).part_2())  # 317

File: year_2023/day_10/day10.go
package main

import (
	"fmt"
	"github.com/Nyaaa/advent-of-code/tools"
	"strings"
)

type cart struct {
	location, start, direction complex128
}

func (c *cart) rotate(tile string) {
	switch tile {
	case "L", "7":
		if real(c.direction) == 0 {
			c.direction *= -1i
		} else {
			c.direction *= 1i
		}
	case "J", "F":
		if real(c.direction) == 0 {
			c.direction *= 1i
		} else {
			c.direction *= -1i
		}
	}
}

func getTrack(c cart, pipes map[complex128]string) []complex128 {
	track := []complex128{}

	for {
		track = append(track, c.location)
		c.location += c.direction
		c.rotate(pipes[c.location])

		if c.location == c.start {
			break
		}
	}

	return track
}

func part2(track []complex128) int {
	area := 0.0

	for i := 0; i < len(track); i++ {
		area += (imag(track[i]) * real(track[(i+1)%len(track)]))
		area -= (real(track[i]) * imag(track[(i+1)%len(track)]))
	}

	if area < 0 {
		area = -area
	}

	return int(area/2) - (len(track) / 2) + 1
}

func solve(input string) (int, int) {
	data := tools.ReadLines(input)
	pipes := map[complex128]string{}
	pipeChars := [7]string{"-", "|", "F", "7", "J", "L", "S"}
	cart := cart{0, 0, 0}

	for i, line := range data {
		for j, tile := range strings.Split(line, "") {
			location := complex(float64(i), float64(j))

			for _, char := range pipeChars {
				if char == tile {
					pipes[location] = tile

					if tile == "S" {
						cart.location = location
						cart.start = location
					}
				}
			}

			pipes[location] = tile
		}
	}

	for _, i := range [4]complex128{1, -1, 1i, -1i} {
		// This is potentially fragile if there are adjacent pipes that don't connect to S.
		_, ok := pipes[cart.location+i]

		if ok {
			cart.direction = i

			break
		}
	}

	track := getTrack(cart, pipes)

	return len(track) / 2, part2(track)
}

func main() {
	fmt.Println(solve("test3.txt"))
	fmt.Println(solve(tools.GetData(2023, 10)))
}
